### DISCUSSION TITLE
metals-emacs does not seem to work with eglot

### DISCUSSION THREAD

**Discussion by moberer** [joaotavora/eglot#934] [Q&A] [answered]
Hi!

I'm trying to use eglot for Scala-development and have run into an issue.
My problem is as follows:

I open a file in the project.
The eglot successfully connects to the Metals server.
If I enter a build.sbt file, I get prompted (as I should be) whether I want to import the build.
I accept and give it a moment.
However, the build import seems to fail and I only get file-local completions.

The reason I'm asking this here and not with Scalas metals is, that the setup works flawlessly when I'm using `lsp-mode`/`metals-lsp`.
This makes me think that my Scala / Metals setup is ok.
Hence my thought is that there is some subtle incompatibility here.

Setup:
metals-emacs (0.11.2)
Doom Emacs (latest)
Emacs 28.1
Java 17.0.3
openSUSE Tumbleweed

Is there any other person successfully using eglot with metals-emacs?
Does anyone have an idea what is wrong here?

Thanks a lot!

Attached: 
- Output from eglot-events-buffer;  [eglot-events.txt](https://github.com/joaotavora/eglot/files/8564631/eglot-events.txt)
- Output from metals-doctor (the html bit included in the eglot-events file)
[metals-doctor](https://user-images.githubusercontent.com/58148625/165337543-2b2fbf9a-6af4-4f70-8411-e3396a21318c.png)
- Log file from metals [metals.log](https://github.com/joaotavora/eglot/files/8564654/metals.log) (Note that the included exception seems not to be the problem; It also happens with lsp-mode)

**Comment by mdbergmann** [comment]
I can't speak for sbt. But I'm using eglot successfully with a Gradle project. The 'import project' option is in fact coming from Metals (IMO, or even Bloop) which does the magic. For Gradle that is just a Gradle wrapper that applies the 'bloopInstall' task.
Have you tried creating the Bloop project manually?

**Reply by moberer** [reply]
Ok, it's reassuring that it works with gradle.
No, I haven't created the Bloop project manually.

I did so now according to [https://scalacenter.github.io/bloop/setup#sbt](https://scalacenter.github.io/bloop/setup#sbt), but got stuck on this:
```
sbt:SWP Interpreter> bloopInstall
[warn] 
[warn] 	Note: Some unresolved dependencies have extra attributes.  Check that these dependencies exist with the requested attributes.
[warn] 		ch.epfl.scala:sbt-bloop:1.5.0 (sbtVersion=1.0, scalaVersion=2.13)
[warn] 
[warn] 	Note: Unresolved dependencies path:
[error] Couldn't run bloopGenerate for aexp-interpreter
[error] Couldn't run bloopGenerate for aexp-interpreter-test
[success] Total time: 2 s, completed 26.04.2022, 17:52:26
```

I'm not entirely sure where this is what it is complaining about.

**Comment by mdbergmann** [comment]
Can't tell.
But I'd assume once the Bloop project is there eglot (Metals) can work with it.

**Comment by moberer** [comment]
I have now tried using VScodes Scala/Metals plugin to init the .bloob folder. 
Metals works as expected in VSCode.
In Emacs this *somewhat* worked. 

I now *do* get a working jump-to-reference on *some* functions/types..
I'm presuming the ones, which the VSCode Metals already computed and cached. 
However, most of them are still broken.

(I'm also prompted to re-import the build, but I have to deny that to have even the select few symbols working).

This seems to point to the fact that, somehow, the metals server is not generating new information while connected with eglot. But it is still able to access previously-generated information.

Frankly, I'm at a loss what is going on there.

**Comment by mdbergmann** [comment, marked as answer]
Regarding jump-to-reference, what I found was that in eglot on a large project, it takes a few minutes until all the diagnostics have been received. Only then the reference jump will work.
You can check in the eglot buffer, there is usually a lot of activity. Once it gets silent all diagnostics have been received.
But I'm not sure how it is actually done. Is the jump-to-reference being done in Metals, or eglot?

**Reply by moberer** [reply]
I've finally found the error:
I was being idiotic, installed an old version of metals-emacs (due to an outdated setup guide) and hence it didn't work.
(Didn't use the official guide, because their coursier setup didn't work for me).

Now it works flawlessly.

Anyways, many thanks for your help!
